core/src/main/java/org/javarosa/form/api/FormEntrySession.java

```diff
@@ -0,0 +1,214 @@
+        }
+
+        return formEntrySession;
+    }
+
+    public static class FormEntryAction implements Serializable {
+        public final String formIndexString;
+        public final String value;
+        public final String action;
+        private final static String NEW_REPEAT = "NEW_REPEAT";
+        private final static String SKIP = "SKIP";
+        private final static String VALUE = "VALUE";
+
+        private FormEntryAction(String formIndexString, String value, String action) {
+            this.formIndexString = formIndexString;
+            this.value = value;
+            this.action = action;
+        }
+
+        public static FormEntryAction buildNewRepeatAction(String formIndexString) {
+            return new FormEntryAction(formIndexString, "", NEW_REPEAT);
+        }
+
+        public static FormEntryAction buildValueSetAction(String formIndexString, String value) {
+            return new FormEntryAction(formIndexString, value, VALUE);
+        }
+
+        public static FormEntryAction buildSkipAction(String formIndexString) {
+            return new FormEntryAction(formIndexString, "", SKIP);
+        }
+
+        public static FormEntryAction buildNullAction() {
+            return new FormEntryAction("", "", "");
+        }
+
+        @Override
+        public String toString() {
+            if (NEW_REPEAT.equals(action)) {
+                return "((" + formIndexString + ") (NEW_REPEAT))";
+            } else if (VALUE.equals(action)) {
+                // TODO PLM: escape 'value' field
+                return "((" + formIndexString + ") (VALUE) (" + value + "))";
+            } else if (SKIP.equals(action)) {
+                return "((" + formIndexString + ") (SKIP))";
+            } else {
+                return "";
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return formIndexString.hashCode() ^ value.hashCode() ^ action.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) {
+                return true;
+            }
+            if (other instanceof FormEntryAction) {
+                FormEntryAction otherAction = (FormEntryAction)other;
+                return formIndexString.equals(otherAction.formIndexString) &&
+                        value.equals(otherAction.value) &&
+                        action.equals(otherAction.action);
+            }
+            return false;
+        }
+
+        public static FormEntryAction fromString(String entryActionString) {
+            String unwrappedEntryActionString =
+                    entryActionString.substring(1, entryActionString.length() - 1);
+            Vector<String> actionEntries = splitTopParens(unwrappedEntryActionString);
+            int entryCount = actionEntries.size();
+
+            if (entryCount != 2 && entryCount != 3) {
+                throw new RuntimeException();
```

Comment:
I added a message. Don't really want to add comprehensive error handling at this point because I'm the only one using this. Currently when this `RuntimeException` is raised it crashes the app